viz/cmd/check.go

```diff
@@ -0,0 +1,282 @@
+				if kerrors.IsNotFound(err) {
+					cert, err = hc.FetchCredsFromOldSecret(ctx, vizNamespace, tapOldTLSSecretName)
+				}
+				if err != nil {
+					return err
+				}
+				return hc.CheckCertAndAnchorsExpiringSoon(cert)
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("tap api service is running").
+			WithHintAnchor("l5d-tap-api").
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return hc.CheckAPIService(ctx, linkerdTapAPIServiceName)
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("viz extension pods are running").
+			WithHintAnchor("l5d-viz-pods-running").
+			Warning().
+			WithRetryDeadline(hc.RetryDeadline).
+			SurfaceErrorOnRetry().
+			WithCheck(func(ctx context.Context) error {
+				pods, err := hc.KubeAPIClient().GetPodsByNamespace(ctx, vizNamespace)
+				if err != nil {
+					return err
+				}
+
+				// Check for relevant pods to be present
+				err = healthcheck.CheckForPods(pods, []string{"linkerd-grafana", "linkerd-prometheus", "linkerd-web", "linkerd-tap"})
+				if err != nil {
+					return err
+				}
+
+				return healthcheck.CheckPodsRunning(pods)
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("linkerd-viz pods are injected").
+			WithHintAnchor("l5d-viz-pods-injection").
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				pods, err := hc.KubeAPIClient().GetPodsByNamespace(ctx, vizNamespace)
+				if err != nil {
+					return err
+				}
+				return healthcheck.CheckIfDataPlanePodsExist(pods)
+			}))
+
+	// TODO: Add dataplane metrics in prometheus check
+	return healthcheck.NewCategory(linkerdVizExtensionCheck, checkers, true)
+}
+
+func newCheckOptions() *checkOptions {
+	return &checkOptions{
+		wait:   300 * time.Second,
+		output: healthcheck.TableOutput,
+	}
+}
+
+func (options *checkOptions) validate() error {
+	if options.output != healthcheck.TableOutput && options.output != healthcheck.JSONOutput {
+		return fmt.Errorf("Invalid output type '%s'. Supported output types are: %s, %s", options.output, healthcheck.JSONOutput, healthcheck.TableOutput)
+	}
+	return nil
+}
+
+func newCmdCheck() *cobra.Command {
+	options := newCheckOptions()
+	cmd := &cobra.Command{
+		Use:   "check [flags]",
+		Args:  cobra.NoArgs,
+		Short: "Check the Linkerd Viz extension for potential problems",
+		Long: `Check the Linkerd Viz extension for potential problems.
+
+The check command will perform a series of checks to validate that the Linkerd Viz
+extension is configured correctly. If the command encounters a failure it will
+print additional information about the failure and exit with a non-zero exit
+code.`,
+		Example: `  # Check that the viz extension is up and running
+  linkerd viz check`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			// Get viz Extension Namespace
+			ns, err := getNamespaceOfExtension(vizExtensionName)
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err.Error())
+				os.Exit(1)
+			}
```

Comment:
This is redundant with the "linkerd-viz Namespace exists"  check. That check should set a `hc.VizNamespace` as a side-effect for the other checks to use (I know it's a bit ugly to use a side-effect, but let's keep it so till we figure a better general framework).

Reply:
That side-effect logic makes sense but adding `vizNamespace` under `hc` could be unnecessary. I can make the check update this global `vizNamespace` variable. wdyt?

Reply:
I'm actually using such a `hc.vizNamespace` field to retrieve a viz api client in my public-api split work.

Reply:
Hmmm. I'm okay with adding that `vizNamespace in `hc`. I can do that in this PR if there is no cocern. WDYT? @adleong

Reply:
I'm not dogmatic but here are my priorities in order:

1. The check command should be performant.  If re-creating a bunch of API clients make the command feel sluggish then we shouldn't do that.  If it has no noticeable performance impact, then let's not bend over backwards to be efficient
2. We should be consistent.  Let's follow the same patterns across the different check commands
3. Side-effects can sometimes be harder to reason about.  Let's avoid them if possible, but not at the expense of the above 2 priorities.

In this case, specifically, I think the redundancy of the namespace check is okay.  Since the first namespace check is fatal, we would expect that if we even get to this check, the namespace lookup should never fail.  (of course, we should still return the error if we somehow encounter it here, just to be defensive).

Reply:
Hmm, The current update i.e using `hc.nam|espace` with only a single namespace check that is fatal does seem to follow the priorities (especially the first two) and would also unblock Alejandros work.